### DISCUSSION TITLE
How to display a dialog when the application is closing?

### DISCUSSION THREAD

**Discussion by harvinders** [microsoft/microsoft-ui-xaml#8459] [Q&A] [answered]
I am trying to display a dialog to confirm if the user would like to save their work before the application is closed, but it does not seem to be working, either the application crashes or it hangs.

I am trying to subscribe to the `AppWindow.Closing` event and display a content dialog using `ContentDialog.ShowAsync()`. 
* `ContentDialog.ShowAsync().GetAwaiter().GetResult()` cause Unhandled exception. 
* I can also not see any `Deferral` documentation that could help pause the app close whilst I display the dialog. 

 I suppose this a very basic ask. Am I missing something or WinUI3/WinAppSDK missing this feature too?

**Comment by jp-weber** [comment]
In the AppWindow.Closing event, I would interrupt the closing ```e.Handled = true``` and then open the dialog in another task. In the dialog task, I would suggest setting a boolean or an enum at this point for a status whether the user confirms the closing. if this is confirmed, the status must be set and the application closed via the exit command. In this case, of course, ```AppWindow.Closing``` is called up again, in which case the closing of the application must not be prevented on the basis of the status.

I have implemented something similar in an application that interacts with a trayicon.

**Comment by castorix** [comment, marked as answer]
This works for me (Windows 10 22H2, Windows app SDK 1.1.0) :

        private async void AppW_Closing(Microsoft.UI.Windowing.AppWindow sender, Microsoft.UI.Windowing.AppWindowClosingEventArgs args)
        {
            args.Cancel = true;

            ContentDialog cd = new ContentDialog
            {
                Title = "Question",
                Content = "Do you want to quit ?",
                PrimaryButtonText = "Yes",
                SecondaryButtonText = "No"
            };
            cd.XamlRoot = this.Content.XamlRoot;
            var cdResult = await cd.ShowAsync();
            if (cdResult == ContentDialogResult.Primary)
            {
                this.Close();
            }
        }

**Comment by gluetheknot2** [comment]
WHERE DO I ADD THIS????!!!!!

**Reply by DarranRowe** [reply]
The MainWindow class.
This is the only place where you can obtain the AppWindow to subscribe to the Closing event, which the answer is the handler for. This is the only place where you can also use the Content property to obtain the XamlRoot from.
